presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
+            }
+            return new Page(1, finalResults.stream().toArray(Block[]::new));
+        }
+    }
+
+    private static class IntermediateLayout
+    {
+        private static final Block FALSE_SINGLE_VALUE_BLOCK;
+
+        static {
+            BlockBuilder falseBlockBuilder = BOOLEAN.createBlockBuilder(null, 1);
+            BOOLEAN.writeBoolean(falseBlockBuilder, false);
+            FALSE_SINGLE_VALUE_BLOCK = falseBlockBuilder.build();
+        }
+
+        private final int intermediateIsDistinctCount;
+        private final List<Boolean> intermediateIsDistinct;
+        private final List<Type> intermediateTypes;
+        private final List<AggregationInputChannel> inputs;
+        private final List<Block> inputSingleNullBlocks;
+
+        public IntermediateLayout(List<InternalAccumulatorFactory> accumulatorFactories, List<AggregationInputChannel> inputs)
+        {
+            requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+            requireNonNull(inputs, "inputs is null");
+            this.intermediateIsDistinctCount = (int) accumulatorFactories.stream()
+                    .filter(InternalAccumulatorFactory::isDistinct)
+                    .count();
+            this.intermediateIsDistinct = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::isDistinct)
+                    .collect(toImmutableList());
+            this.intermediateTypes = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::getIntermediateType)
+                    .collect(toImmutableList());
+
+            this.inputs = ImmutableList.copyOf(requireNonNull(inputs, "inputs is null"));
+            this.inputSingleNullBlocks = accumulatorFactories.stream()
+                    .map(accumulatorFactory -> accumulatorFactory.getIntermediateType().createBlockBuilder(null, 1)
+                            .appendNull()
+                            .build())
+                    .collect(toImmutableList());
+        }
+
+        private int getOutputChannelCount()
+        {
+            return 1 + intermediateTypes.size() + inputs.size();
+        }
+
+        public boolean hasDistinct()
+        {
+            return intermediateIsDistinctCount > 0;
+        }
+
+        public boolean hasNonDistinct()
+        {
+            return intermediateIsDistinctCount < intermediateIsDistinct.size();
         }
 
-        public void processPage(Page page)
+        private int getIntermediateChannel(int accumulatorIndex)
         {
-            if (step.isInputRaw()) {
-                aggregation.addInput(page);
+            checkElementIndex(accumulatorIndex, intermediateTypes.size());
+            return 1 + accumulatorIndex;
+        }
+
+        private Optional<Page> buildPartialDistinctRows(Page page, List<Optional<Block>> partialDistinctMasks)
+        {
+            checkArgument(partialDistinctMasks.size() == intermediateTypes.size());
```

Comment:
The callers of this method seem to be adding items to `partialDistinctMasks` only for distinct aggregations, hence, it can be that `partialDistinctMasks.size() < intermediateTypes.size()`, but `partialDistinctMasks.size() == intermediateIsDistinctCount` should hold.